feat(lint): deno lint --json

souldzin opened the pull request:
## Description

This PR implements a `--json` option for `deno lint` to print the result in JSON format 

**Heads up!** This is @JackSkylark and my first time working with Rust so apologies in advanced for any trivial oversights 😅 

## Example

```
$ deno lint --unstable --json test.ts
{
  "diagnostics": [
    {
      "location": {
        "filename": "/path/to/file.ts",
        "line": 3,
        "col": 4
      },
      "message": "assign `this` to declare a value is not allowed",
      "code": "no-this-alias",
      "line_src": "    const self = this;",
      "snippet_length": 18
    }
  ],
  "runtime_errors": []
}
```

## References

- https://github.com/denoland/deno/issues/6721

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno?pullRequest=6940) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno?pullRequest=6940) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: souldzin<br/>:x: JackSkylark<br/><hr/>**souldzin** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno?pullRequest=6940) it.</sub>

souldzin commented on cli/flags.rs:
```diff
@@ -56,6 +56,8 @@ pub enum DenoSubcommand {
     files: Vec<String>,
     rules: bool,
+    json: bool,
+    json_sort: bool,
```
**What's up with `json_sort`?**

This was needed to enforce determinism with the integration spec. `json_sort` is a flag sent to the `JsonLintReporter` to sort the lints by location before serialization.

lucacasonato replied:
I am for just always sorting (maybe even in the pretty printer too?). @bartlomieju WDYT?

bartlomieju replied:
Agree, `--sort` is superfluous

souldzin replied:
> I am for just always sorting (maybe even in the pretty printer too?)

Sounds good to me! Since JSON has to wait tell the whole object is ready anyways, the cost of just always sorting is fairly small :smile:

IMHO, we shouldn't sort in the `PrettyLintReporter`. If we wanted to do this we'd either need to store all diagnostics and wait to `eprintln` tell the very end, or introduce some complexity such that we're guaranteed that the file's are visited and resolved in a deterministic order. For the pretty printer, I'm not sure sorting is worth it 🤔

It's really nice that the default pretty reporter is responsive (i.e., it outputs results right away) and has very low overhead (i.e., it isn't burdened with remembering diagnostics). Most use cases I can think of for wanting to run `deno lint`, I think responsiveness and low overhead are most important. 🤔

Maybe we should create a follow-up issue about sorting the result of the `PrettyLintReporter` and continue the discussion there?

souldzin commented on cli/flags.rs:
```diff
@@ -1014,6 +1026,19 @@ Ignore linting a file by adding an ignore comment at the top of the file:
+      Arg::with_name("sort")
+        .long("sort")
+        .help("Sort the lints of the outputted JSON.")
+        .requires("json")
```
**note:** Limiting the use of `--sort` to `--json` because we don't want or need to sort with the console pretty print reporter.

souldzin commented on cli/lint.rs:
```diff
@@ -21,46 +19,50 @@ use deno_lint::rules;
   let target_files = collect_files(args)?;
   debug!("Found {} files", target_files.len());
 
-  let error_count = Arc::new(AtomicUsize::new(0));
```
**Why is this changing to `has_error`?**

Keeping track of the number of error's is now a responsibility of the "reporter". The only thing we need to know is whether or not to `exit(1)` so we just keep track of if we have an error or not.

You'll notice that this fixes a preexisting bug where the lint resulted in a `Err` but we did not `exit(1)` because the the `error_count` was `0`.

souldzin commented on cli/lint_reporter.rs:
```diff
@@ -0,0 +1,125 @@
+use serde::Serialize;
+
+pub enum LintReporterKind {
+  Console,
```
**thought:** I realize now that this should probably be called `Pretty` 😅 What do you think?

JackSkylark replied:
Agreed. At the moment all linting writes to the console. Also, this might be confusing if we ever want to add different output targets.

souldzin replied:
Done! Renamed `ConsoleX` to `PrettyX`

souldzin commented on cli/lint_serialize.rs:
```diff
@@ -0,0 +1,46 @@
+use serde::Serialize;
+
+#[derive(Serialize)]
+struct SerializableLocation {
```
**What's going on here?**

Since the `LintDiagnostic` struct does not implement the `Serialize` trait, **and** it's defined in a separate crate, we need to use adapters to make create a "serializable" version of it. These `SerializableX` structs are simply adapters to the `deno_lint` structs, that also `derive(Serialize)`.

lucacasonato replied:
Makes sense. Because `deno_lint` is part of Deno (https://github.com/denoland/deno_lint), we could just make the `deno_lint::diagnostic::LintDiagnostic`  serialisable upstream so we don't need this wrapper. Would save a minuscule amount of memory and CPU and would simplify this a little. @bartlomieju WDYT?

bartlomieju replied:
`LintDiagnostic` should already be JSON-serializable, if it's not, then we should make it so in `deno_lint`

souldzin replied:
> `LintDiagnostic` should already be JSON-serializable

Yeah, unfortunately it doesn't look like this is the case. 😞 

```
error[E0277]: the trait bound `deno_lint::diagnostic::LintDiagnostic: serde::Serialize` is not satisfied
   --> cli/lint.rs:242:3
    |
242 |   lints: Vec<LintDiagnostic>,
    |   ^^^^^ the trait `serde::Serialize` is not implemented for `deno_lint::diagnostic::LintDiagnostic`\
```

Also `deno_lint` doesn't have `serde` as a dependency, so I'm not sure how eager we are to add one just for this :shrug:

Having `deno_lint::LintDiagnostic` serializable and using that here would be the simplest, but it's important to be aware of the tradeoff. If we made `LintDiagnostic` serializable and just used that, then a change to the struct in `LintDiagnostic` would change the API of `deno lint`. Thankfully we have an integration test covering the expected API, but that test is in a separate project, meaning that it would be fairly easy to merge changes to `deno_lint` without realizing they would "break" things (hopefully just pipelines).

- Are we sure we like the simplicity over control tradeoff here?
- Is this something we want to take care of now or iteratively in a follow-up?

Completely your call what you want to do here. Just let me know :smile:

souldzin replied:
> would save a minuscule amount of memory

@lucacasonato FWIW, I'm not entirely sure we'd save memory because we'd still need to clone the `LintDiagnostic` before pushing it to the `Vec` 🤔

I'm not super familiar with managing the life cycle of these things though. Please let me know if I'm missing something 😅

bartlomieju replied:
@souldzin it's okay to do that in `deno_lint`. I've opened an issue: https://github.com/denoland/deno_lint/issues/253 Will try to get it done soon

souldzin replied:
> I've opened an issue: denoland/deno_lint#253 Will try to get it done soon

Thanks @bartlomieju! Sounds good to me :smile:

If you are not able to get to it, I might be able to jump on it as well. Just let me know :smile:

bartlomieju replied:
@souldzin if you want to work on it, then go ahead, I described details in the issue.

souldzin replied:
Thanks @bartlomieju for the upstream fixed. Rebased with `master` and removed the `Serializable` wrappers 👍

souldzin commented on cli/tests/integration_tests.rs:
```diff
@@ -2208,6 +2208,13 @@ itest!(deno_lint {
+    "lint --unstable --json --sort lint/file1.js lint/file2.ts lint/ignored_file.ts",
+  output: "lint/expected_json.out",
+  exit_code: 1,
+});
```
**question:** This integration test gives us some good coverage, but it would be nice to have some unit tests for some of the modules. Would we like to see those in this PR or in a follow-up? I'm very new to rust, so some examples of this would be much appreciated :smile:

bartlomieju replied:
We already test standard library for clean lint in `std_lint` test

souldzin replied:
> We already test standard library for clean lint in std_lint test

Right. It seemed like the core functionality had a good amount of test coverage. I felt weird submitting a PR with no *unit tests*, but most of this is "Glue code", so an *integration test* is probably the most important 😅 

I'll go ahead and resolve this.

souldzin commented:
Here's the PR @JackSkylark from our pairing session. Would you mind taking a look at it before I pass it along to a project maintianer? Thanks! :bow:

souldzin commented:
@JackSkylark oh! You'll also need to sign the CLA 😅

lucacasonato commented on cli/lint_reporter.rs:
```diff
@@ -0,0 +1,125 @@
+
+#[derive(Serialize)]
+struct JsonLintReporter {
+  lints: Vec<SerializableLint>,
```
I think we should use `diagnostics` rather than `lints` because that is what these are called in deno_lint and TypeScript.

So:
- `lints` -> `diagnostics`
- `SerializableLint` -> `SerializableDiagnostic`

bartlomieju replied:
Agreed on `lints` -> `diagnostics`, but I feel that `SerializableLint` is superfluous... `deno_lint::Diagnostic` is already JSON-serializable, so it should be only a matter of stringifying it

bartlomieju commented on cli/flags.rs:
```diff
@@ -990,6 +999,9 @@ fn lint_subcommand<'a, 'b>() -> App<'a, 'b> {
   deno lint --unstable myfile1.ts myfile2.js
 
+Print result as JSON:
+  doeno lint --unstable --json
```
`s/doeno/deno`

bartlomieju commented on cli/lint_reporter.rs:
```diff
@@ -0,0 +1,125 @@
+#[derive(Serialize)]
+struct JsonLintReporter {
+  lints: Vec<SerializableLint>,
+  runtime_errors: Vec<LintError>,
```
What is this field used for? Linter does only static analysis, so I don't expect to see any runtime errors

souldzin replied:
> What is this field used for?

@bartlomieju this is a good question. Technically, [`lint_file`](https://github.com/denoland/deno/blob/fcea80391d014cf3ae3e15bdf3443fe4488443fb/cli/lint.rs#L65) returns a result object which could hold an `Err` signifying that something went wrong during the linting itself. Previously we'd just `eprintln` the error information (which is what the `PrettyLintReporter` does now), but if we did this with the JSON reporter, we'd break the JSON syntax of the CLI's output. So we needed someway to present this information in the JSON object itself.

I was thinking of calling this `errors`, but I thought that would be easily confused with "lint violations". What do you think? Is `errors` a better name here? Is there another name you think is more appropriate?

bartlomieju replied:
I see, that makes sense. I think we should go with `diagnostics` and `errors`.

souldzin replied:
> I see, that makes sense. I think we should go with diagnostics and errors.

Done! Thanks @bartlomieju!

bartlomieju commented on cli/lint_serialize.rs:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  pub fn get_sort_key(&self) -> String {
+    let location = &self.location;
+
+    return format!("{}:{}:{}", location.filename, location.line, location.col);
+  }
```
Please merge `lint_serialize.rs` and `lint_reporter.rs` into `lint.rs`

bartlomieju commented:
@souldzin `LintDiagnostic` can now be serialized to JSON, you just need to merge latest `master` branch.

souldzin commented:
Looking at this now :eyes:

souldzin commented on cli/lint.rs:
```diff
@@ -161,21 +177,106 @@ fn lint_file(file_path: PathBuf) -> Result<Vec<LintDiagnostic>, ErrBox> {
+      .sort_by(|a, b| get_sort_key(&a).cmp(&get_sort_key(&b)));
+
+    let json = serde_json::to_string_pretty(&self);
+    println!("{}", json.unwrap());
```
**Heads up!** I just updated the `JsonLintReporter` to print to stdout instead of stderr so that we can easily pipe the output `deno lint --unstable --json | pbcopy`. Seemed pretty helpful and I couldn't find a reason why printing to `stderr` would be better in the json case 🤔

nayeemrmn replied:
> I couldn't find a reason why printing to `stderr` would be better in the json case 🤔

Because that's what it does without `--json` and it is much simpler to treat `--json` as strictly a different output format. If you _were_ to change it to stdout, you'd have to change the exit code to be unconditionally 0 since it clearly no longer treats diagnostics as "errors".

souldzin replied:
> If you were to change it to stdout, you'd have to change the exit code to be unconditionally 0 since it clearly no longer treats diagnostics as "errors".

It looks like we've already been doing this https://github.com/denoland/deno/pull/6940/files#diff-8dcab4b639b85d952afd2523e77a0b3dR80. 🤔 

I'm fine either way, but writing to stdout is a nice convenience and seems to be what other linting CLI's do (i.e. `eslint`). We can handle this in a follow-up issue if you'd prefer :shrug:

nayeemrmn replied:
> It looks like we've already been doing this https://github.com/denoland/deno/pull/6940/files#diff-8dcab4b639b85d952afd2523e77a0b3dR80. 🤔

No? The line you linked says we haven't. Since you're not just talking about `--json`, we've already discussed this: #6303. `eslint` is not the only existing linter and probably not the one we want to copy.

Please revert this change.

souldzin replied:
> Please revert this change.

No problem!

> No? The line you linked says we haven't.

Ah! I see. I misunderstood what you were saying in the first place. I haven't interpreted stderr and error codes this way before but I don't have a strong opinion either way. Reverted the commit 👍

souldzin commented:
Thanks @bartlomieju! Rebased and removed the `Serializable` wrappers. Also please note:

- https://github.com/denoland/deno/pull/6940#discussion_r467527541

Back to you! ⚽

bartlomieju commented on cli/main.rs:
```diff
@@ -355,7 +356,13 @@ async fn lint_command(
   }
 
-  lint::lint_files(files).await
+  let reporter = if json {
+    lint::LintReporterKind::Json
+  } else {
+    lint::LintReporterKind::Pretty
+  };
+
+  lint::lint_files(files, reporter).await
```
I guess we should pass `json` argument to `lint_files` - no need to leak `LintReporterKind` outside of `lint` module

bartlomieju commented on cli/tests/lint/expected_json.out:
```diff
@@ -0,0 +1,38 @@
+      "snippet_length": 2
+    }
+  ],
+  "errors": []
```
Nice test, could you add some file that will fill `errors` list as well?

bartlomieju commented:
I've added a few fixes and changed test.

souldzin commented:
> I've added a few fixes and changed test.

Lovely fixes @bartlomieju! I really appreciate everyone's patience and attention to detail. This has been a great contributing experience :smile:
